### PR TITLE
Redo state manager

### PR DESCRIPTION
**Pull request by applmak**
The rewrite has a brand new API. Salient details:
- Now, interpolators are defined on client, where they are used, and are
  no longer needlessly sent across from the server, which doesn't need
  them.
- Also, the server only tracks the most recent data, because it only
  ever sends the most recent data.
- Finally, communication for the state happens on the framework sockets,
  not the individual modules, which should prevent some race conditions
  that happen when client/server instantiation order isn't guaranteed.

### CODE REVIEW

File: client/state/state_manager.js [hunk 1]
```diff
@@ -13,86 +13,107 @@ See the License for the specific language governing permissions and
...
-  set(value, time) {
-    if (this.owner_ !== undefined && this.network_.id !== this.owner_) {
-      throw new Error('Attempted to set state for state owned by ' + this.owner_);
-    }
-    super.set(value, time);
-    this.network_.emit('newclientstateset',
-        { name: this.name_, value: value, time: time });
-  }
+
+function initState() {
```
Comment (jmegq) [inline client/state/state_manager.js]: It smells like State wants to be a class, but this is fine.

File: client/state/state_manager.js [hunk 2]
```diff
@@ -13,86 +13,107 @@ See the License for the specific language governing permissions and
...
+      return {
+        define(stateName, def) {
+          assert(!(stateName in stateMap[id].state), `State ${stateName} was already defined!`);
+          stateMap[id].state[stateName] =
+              new SharedState(stateName, decodeInterpolator(def));
+          if (stateMap[id].priorData[stateName]) {
+            // We have some data that the server sent before we were ready.
+            // Add it to the shared state now.
+            for (const {data, time} of stateMap[id].priorData[stateName]) {
+              stateMap[id].state[stateName].set(data, time);
```
Comment (jmegq) [inline client/state/state_manager.js]: Might help some module writer's sanity to log if we clobber a bit of state with priorState here, i.e. if the data are different but time is the same.

File: demo_modules/balls/balls_client.js [hunk 3]
```diff
@@ -27,18 +27,19 @@ export function load(state, wallGeometry) {
     async willBeShownSoon(container) {
       this.surface = new CanvasSurface(container, wallGeometry);
       this.canvas = this.surface.context;
+
+      this.ballsState = state.define('balls', [{
+        x: 'NumberLerpInterpolator',
```
Comment (jmegq) [inline demo_modules/balls/balls_client.js]: Something like Interpolator.NumberLerp feels nicer than magic string constant.

File: client/modules/module.js [hunk 4]
```diff
@@ -228,6 +229,7 @@ export class ClientModule {
     moduleTicker.remove(this.instance);
 
     if (this.network) {
+      this.stateManager.close();
       this.network.close();
```
[the remark this answers is not in the document]
Comment (applmak) [reply client/modules/module.js]: Done.

File: client/state/state_manager.js (hunk 1 above)
Comment (applmak) [reply client/state/state_manager.js]: Good idea. Done.

File: client/state/state_manager.js (hunk 2 above)
Comment (applmak) [reply client/state/state_manager.js]: I'm not 100% sure how to do that, so I added a TODO.

File: demo_modules/balls/balls_client.js (hunk 3 above)
Comment (applmak) [reply demo_modules/balls/balls_client.js]: Ah, good point.

File: client/state/state_manager.js [hunk 5]
```diff
@@ -13,86 +13,108 @@ See the License for the specific language governing permissions and
...
 }
 
-// Describes something that tracks all sharedstate and provides methods for
-// the communication of that state across the network.
-export class StateManager {
-  constructor(network) {
-    // A map of tracked state name -> state variable.
-    this.trackedState_ = {};
-    this.network_ = network;
+function isClosedOrStale(state) {
```
Comment (jmegq) [inline client/state/state_manager.js]: lol, should be a method on StateRecord, no?

File: demo_modules/gears/gears_client.js [hunk 6]
```diff
@@ -19,6 +19,19 @@ export function load(debug, state, wallGeometry) {
 
       // A map of teeth -> Path2D.
       this.gearPaths_ = [];
+
+      this.gearsState = state.define('gears', {
+        x: a => a,
```
Comment (jmegq) [inline demo_modules/gears/gears_client.js]: So no interpolators?

File: client/state/state_manager.js (hunk 5 above)
Comment (applmak) [reply client/state/state_manager.js]: Oops, yeah.

File: demo_modules/gears/gears_client.js (hunk 6 above)
Comment (applmak) [reply demo_modules/gears/gears_client.js]: There weren't any here originally. Gears doesn't really need the 'state' abstraction; it was just a more reliable way to do the networking.

**Pull request merged**